### PR TITLE
Multisig-wallet ETH signature support

### LINKED ISSUE #10: Supports both Ethereum & Fuel signature recovery
Similar to the [EIP-191 Signed Data Standard](https://eips.ethereum.org/EIPS/eip-191) the wallet should support recovery for both 

- [ ] #13
- [ ] #14

### PR DESCRIPTION
**Pull request by K1-R1**
## Type of change

- New feature

## Changes

- This PR adds the ability for both Fuel and EVM addresses to be owners/users of the multisig-wallet; by enabling signature recovery from signatures over EIP-191 formatted and Ethereum prefixed message hashes.
- It additionally adds testing and updated documentation.

## Related Issues

Closes #10

### CODE REVIEW

File: eth-account-abstraction/src/utils.sw
Comment (adlerjohn) [inline eth-account-abstraction/src/utils.sw:18]: Not sure this will work correctly. See https://github.com/FuelLabs/sway/issues/2594#issuecomment-1221205410 for more context. It might be worth adding a bytes library as a requirement to implement this correctly.

Comment (K1-R1) [reply eth-account-abstraction/src/utils.sw:18]: Ah, I think I see; so the library would need to give an unpadded reference type for correct hashing with sha256?

**Comment by K1-R1** [comment]
This PR additionally closes the following issues:
- Closes #7 
- Closes #11 
- Closes #12 
- Closes #13 
- Closes #14 

I have not included these in the description as to make reviewing easier.

**Comment by Braqzen** [comment]
> This PR additionally closes the following issues:
> 
> * Closes [Full test coverage #7](https://github.com/FuelLabs/sway-applications/issues/7)
> * Closes [EIP-191 signature support #11](https://github.com/FuelLabs/sway-applications/issues/11)
> * Closes [Full test coverage #12](https://github.com/FuelLabs/sway-applications/issues/12)
> * Closes [Ethereum Signatures #13](https://github.com/FuelLabs/sway-applications/issues/13)
> * Closes [Fuel Signatures #14](https://github.com/FuelLabs/sway-applications/issues/14)
> 
> I have not included these in the description as to make reviewing easier.

I'll review this sometime tomorrow. Note that if you do not add these into the description they will not be closed once this PR is merged.

**Comment by K1-R1** [comment]
No worries, I thought it'd be easier to review like this and then I'll manually close the relevant issues after

File: multisig-wallet/project/multisig-contract/src/main.sw
Comment (matt-user) [inline multisig-wallet/project/multisig-contract/src/main.sw:15]: this should be included in the prelude so no need to import

File: multisig-wallet/project/multisig-contract/src/main.sw
Comment (matt-user) [inline multisig-wallet/project/multisig-contract/src/main.sw:17]: included in the prelude no need to import

File: multisig-wallet/project/multisig-contract/src/main.sw
Comment (matt-user) [inline multisig-wallet/project/multisig-contract/src/main.sw:31]:
```suggestion
    /// The number of approvals required in order to execute a Tx.
```

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (matt-user) [inline multisig-wallet/project/multisig-contract/src/utils.sw:13]: no need to import included in prelude

**Review by matt-user — CHANGES_REQUESTED** [review]
Just took a quick look at the code (haven't looked at tests yet).  One addition I would make to the docs is to link EIP 191 somewhere in them as it is mentioned frequently in the code.  Also I don't see the spec in here, but I may have missed it.

**Review by Braqzen — CHANGES_REQUESTED** [review]
1. SPECIFICATION.md?
2. Imports from the [prelude](https://fuellabs.github.io/sway/v0.31.1/introduction/standard_library.html#standard-library-prelude) should not be explicitly imported because they are automatically imported for us. Some imports have been pointed out in the previous review but not all files
3. When importing a single component from somewhere it should not be wrapped inside `{}` e.g. `utils.sw` for `erc`
4. The interface documentation references `Panics` in a few areas rather than `Reverts`
5. `mod.rs` uses `unsafe {}` blocks. Given that it comes from that is likely well tested it's probably fine; however, this may be problematic for maintenance because you're saying "I have checked this, trust me". Is there a workaround to not use `unsafe` code?
6. There are comments in `mod.rs` function `format_and_sign` that do not insert a space between the `//` and text proceeding e.g. `//Format` instead of `// Format`. That aside, those comments do not convey any insight into what is going on. It's clear enough from the code to see that something is formatted etc. Those comments should be removed or elaborated upon. This may apply to other files as well hence I'm writing it here instead of that specific section.
7. There are no tests for `execute_transaction`

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (nfurfaro) [reply multisig-wallet/project/multisig-contract/src/utils.sw:49]: If `ec_recover_address` fails, it already propagates an `EcRecoverError::UnrecoverablePublicKey` enum, so we could just use that here.

**Review by nfurfaro — COMMENTED** [review]
Looking really good! So nice to see this feature on the brink of being merged. 
I left a few comments, mostly trivial stuff but one which may reduce the number of storage reads being used.

File: multisig-wallet/project/multisig-contract/src/main.sw
Comment (nfurfaro) [reply multisig-wallet/project/multisig-contract/src/main.sw:48]: Ah, ignore that last comment, I see that this is not doing what I initially thought and is not reading from storage at all.

File: multisig-wallet/project/multisig-contract/src/data_structures.sw
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/data_structures.sw:13]: The `None` variant is used to specify that a message isn't formatted/prefixed. This allows `ec_recovery` from a 
variety of scenarios:

-  A signature from a Fuel wallet signing over a message that _was_ [EIP-191](https://eips.ethereum.org/EIPS/eip-191) formatted but _didn't_ have the Ethereum prefix applied.
- A signature from a Fuel wallet signing over a message that _wasn't_ formatted and _didn't_ have the Ethereum prefix applied.
- A signature from an EVM wallet signing over a message that _was_ [EIP-191](https://eips.ethereum.org/EIPS/eip-191) formatted and _had_ the Ethereum prefix applied.

These enums are used for the control flow in `utils.sw::recover_signer`.

File: multisig-wallet/project/multisig-contract/src/interface.sw
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/interface.sw:32]: I see, I agree on the parameters being changed to `signatures`, but I think the struct being called `Signature` is also a bit confusing as it has the field `signature`.
Perhaps `SignatureInfo` for the struct?

File: multisig-wallet/project/multisig-contract/src/interface.sw
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/interface.sw:14]: Backticks are back

File: multisig-wallet/project/multisig-contract/src/interface.sw
Comment (Braqzen) [reply multisig-wallet/project/multisig-contract/src/interface.sw]: Sure

File: multisig-wallet/project/multisig-contract/src/data_structures.sw
Comment (Braqzen) [reply multisig-wallet/project/multisig-contract/src/data_structures.sw:49]: When I was writing the initial documentation I believe the default multisig wallet implementation would not use weighting and thus the default is technically 1, 1 per signature. I may have been incorrect.

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (Braqzen) [reply multisig-wallet/project/multisig-contract/src/utils.sw:58]: Yes, on the one hand that may save memory per call but on the other hand it brings them into global scope where everything has access. I _assume_ that `const` is better for runtime cost.

File: multisig-wallet/project/multisig-contract/src/data_structures.sw
Comment (Braqzen) [reply multisig-wallet/project/multisig-contract/src/data_structures.sw:47]: This is something that I've noticed but it slipped my mind. Looking at the code it's been changed to save on some conversions so it should make it slightly cheaper to use and possibly deploy. The `Address` is correct thought because it prevents someone from accidentally using the `b256` of a contract as a user and then finding out that they cannot sign etc. For clarity and correctness, even at a slight cost, `Address` should be used.

File: multisig-wallet/project/multisig-contract/src/main.sw
Comment (simonr0204) [inline multisig-wallet/project/multisig-contract/src/main.sw:56]: Should we also check that transactions are being executed in nonce order? There are scenarios where a signer would want to approve a transaction on the condition that all previous transactions in the queue have already been executed. Many multisig wallets (e.g. Gnosis Safe) have this requirement.

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (simonr0204) [reply multisig-wallet/project/multisig-contract/src/utils.sw:104]: hmm yes I think this should be keccak.

File: multisig-wallet/project/multisig-contract/src/main.sw
Comment (simonr0204) [inline multisig-wallet/project/multisig-contract/src/main.sw:46]: Should check that threshold <= sum of weights, otherwise wallet can be initialized in an unusable state.

**Comment by simonr0204** [comment]
I think this would be the object of a future PR, but we should consider whether there's functionality here that should be lifted into `std::vm::evm` . Particularly the EIP191 prefix stuff.

File: multisig-wallet/project/multisig-contract/src/main.sw
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/main.sw:56]: As it stands, I don't think it's possible for transactions to be executed in anything but nonce order; the user never inputs a nonce value to either `transfer` or `execute_transaction`, the value is only ever called from storage when executing a transaction. 
A function to cancel an order ( by incrementing the nonce ) would be helpful; I'll add that in

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/utils.sw:58]: 👍

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/utils.sw:60]: I've changed these to `encode_and_pack_signed_data` and `signed_data`, as the `EIP-191` docs refer to this data as `signed_data`.
Open to suggestions if this is still unclear.

File: multisig-wallet/project/multisig-contract/src/main.sw
Comment (simonr0204) [reply multisig-wallet/project/multisig-contract/src/main.sw:61]: That makes sense. It's a feature that's rarely used in practice, but in a scenario where the signers don't all trust each other, being able to explicitly render a previously shared signature useless (by using its nonce) is necessary.

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/utils.sw:101]: Yes

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/utils.sw:104]: No; EIP-191 isn't and nor is the prefix, but the prefix originates from Geth as I understand it.

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/utils.sw:109]: Yh you two are correct; both ethers-rs and go-ethereum use keccak256. Looks like the older references I was using are just outdated or wrong. Will switch to keccack256

File: multisig-wallet/project/multisig-contract/src/utils.sw
Comment (simonr0204) [reply multisig-wallet/project/multisig-contract/src/utils.sw]: Presumably there are other ethereum clients that implement this signature scheme though? I think the prefixes are a standard across implementations. It may be that the client usually attaches the prefix but it's not necessarily the case. One could prepare a signature without using any client software if they follow the spec.

Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/src/utils.sw]: Yh; this scheme is common across clients and wallets, but can be followed manually with the spec. I meant that as the EIP-191 document itself states:

Additionally, 0x19 has been chosen because since ethereum/go-ethereum#2940 , the following is prepended before hashing in personal_sign:
`"\x19Ethereum Signed Message:\n" + len(message)`

It seemed that it's use, with regards to EIP-191, stems from it's integration into Geth

File: multisig-wallet/project/multisig-contract/tests/functions/transaction_hash.rs
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/tests/functions/transaction_hash.rs:52]: I'm unaware of another way to match Sway's encoding without just hardcoding a `u8` array. I can look more deeply into this; opening an issue on the SDK and if a simpler solution arrises I can add it in a future PR. 

Assuming there currently isn't a simpler approach, it's a decent example of manually using the SDK's `ABIEncoder`.

File: multisig-wallet/project/multisig-contract/tests/functions/transfer.rs
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/tests/functions/transfer.rs:87]: I've written a test for unrecoverable key, but I don't think it's panicking for the expected reason. I'd have to make the `ecr` opcode panic, and I couldn't see any examples of this in the Sway repo tests. I'll do some digging on this, but won't add the test until it's definitely working

File: multisig-wallet/project/multisig-contract/tests/utils/mod.rs
Comment (K1-R1) [reply multisig-wallet/project/multisig-contract/tests/utils/mod.rs:132]: I'll add a paths module

**Comment by K1-R1** [comment]
> 5\. `mod.rs` uses `unsafe {}` blocks. Given that it comes from that is likely well tested it's probably fine; however, this may be problematic for maintenance because you're saying "I have checked this, trust me". Is there a workaround to not use `unsafe` code?

Currently theres no workaround, but there is an open issue related to removing the use of unsafe here:
https://github.com/FuelLabs/fuels-rs/issues/698
I'll reference the issue in `mod.rs`

**Comment by K1-R1** [comment]
> 7\. There are no tests for `execute_transaction`

Currently `execute_transaction` doesn't do anything, as it requires the ability to make arbitrary contract calls, which isn't yet supported.

**Comment by Braqzen** [comment]
> > 7. There are no tests for `execute_transaction`
> 
> Currently `execute_transaction` doesn't do anything, as it requires the ability to make arbitrary contract calls, which isn't yet supported.

Template tests should be written and an `#[ignore]` should be added with comments in the tests to describe why we cannot fully test at the moment. If these are not added then someone has to remember that the entire suite is missing. If the tests are partially written, or at least stubbed, then when the tests are run they will come up as ignored and anyone will be able to pick on the tests needing to be written when possible.

**Comment by K1-R1** [comment]
I've bumped the versions on this PR to align with #325, CI will fail until it's merged.

File: multisig-wallet/project/multisig-contract/tests/functions/balance.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/functions/balance.rs:37]: Looking at the imported constant we can see that the default is non-zero; however, this test itself does not guarantee that there is any change in the balance before and after because it does not assert that the initial balance and final balance are not the same. 1 more assertion should be added to check this because someone may change the constant in the other file and this test would still pass.

File: multisig-wallet/project/multisig-contract/tests/functions/cancel_transaction.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/functions/cancel_transaction.rs:24]:
```suggestion
        assert_eq!(final_nonce, initial_nonce + 1);
```

File: multisig-wallet/project/multisig-contract/tests/functions/constructor.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/functions/constructor.rs:56]: I would order the tests to be in the same order as the requirements so that it's easier to see requirement by requirement which test should be tested and thus if anything is missing

File: multisig-wallet/project/multisig-contract/tests/utils/mod.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/utils/mod.rs:172]: Unsure about the name, what about something like `default_users`?

File: multisig-wallet/project/multisig-contract/tests/functions/constructor.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/functions/constructor.rs:84]: This is nice for a one liner; however, you can make the suite more flexible by counting the weight of the users that are being provided into the constructor and then do `weight + 1` for the threshold. This way you can dynamically alter the weights in the setup without worry of brittle tests.

File: multisig-wallet/project/multisig-contract/tests/functions/execute_transaction.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/functions/execute_transaction.rs:14]: I would also add the `expected = "Error"` to make it even clearer for future devs which line should fail.

File: multisig-wallet/project/multisig-contract/tests/functions/nonce.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/functions/nonce.rs:22]:
```suggestion
        assert_eq!(final_nonce, initial_nonce + 1);
```

File: multisig-wallet/project/multisig-contract/tests/functions/transfer.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/functions/transfer.rs:90]: Could use `DEFAULT_TRANSFER_AMOUNT`, no? Also similar comment where the pre-condition is different to the post-condition.

File: multisig-wallet/project/multisig-contract/tests/functions/transfer.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/functions/transfer.rs:105]: Not needed

File: multisig-wallet/project/multisig-contract/src/main.sw
Comment (simonr0204) [reply multisig-wallet/project/multisig-contract/src/main.sw:38]: It's really for "peace of mind" for a signer who has previously signed a transaction (off-chain, but that signed transaction is now out "in the wild" ), and would like to be certain that the transaction can _never_ be executed. So they spend the nonce on-chain.

File: multisig-wallet/project/multisig-contract/tests/utils/mod.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/utils/mod.rs:15]:
```suggestion
abigen!(
    MultiSig,
    "./project/multisig-contract/out/debug/multisig-contract-abi.json"
);
```

File: multisig-wallet/project/multisig-contract/tests/utils/mod.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/utils/mod.rs:29]:
```suggestion
pub struct Caller {
    pub contract: MultiSigContract,
    pub wallet: WalletUnlocked,
}

pub mod paths {
    pub const MULTISIG_CONTRACT_BINARY_PATH: &str = "./out/debug/multisig-contract.bin";
    pub const MULTISIG_CONTRACT_STORAGE_PATH: &str =
        "./out/debug/multisig-contract-storage_slots.json";
}
```

File: multisig-wallet/project/multisig-contract/tests/utils/mod.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/utils/mod.rs:9]:
```suggestion
```

File: multisig-wallet/project/multisig-contract/tests/utils/mod.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/utils/mod.rs:31]:
```suggestion
pub mod interface {
```

File: multisig-wallet/project/multisig-contract/tests/functions/cancel_transaction.rs
Comment (Braqzen) [inline multisig-wallet/project/multisig-contract/tests/functions/cancel_transaction.rs:12]: Can we check that the tests also emit the correct logs? [Example](https://github.com/FuelLabs/sway-applications/pull/322/files#diff-3bf4205ff279244340076b24dce7c26d8ba6c8767c54a9c47566046f3d83c09eR37-R42)

**Review by Braqzen — APPROVED** [review]

**Review by matt-user — COMMENTED** [review]
lgtm, will approve once more people get a chance to review

**Review by Braqzen — APPROVED** [review]

**Review by simonr0204 — APPROVED** [review]

**Pull request merged**
